### DISCUSSION TITLE
UI overhaul/rewrite

### DISCUSSION THREAD

**Discussion by tomangelo2** [colobot/colobot#1813] [General]
Idea: Rework UI code to be more customizable
Goals:
- [ ] UI properly scaling on different aspect ratios
- [ ] Proper support for touchscreens (in case we'd port it to other platforms)
- [ ] Definable by external files, instead of being hardcoded in engine. Possible to enhance/replace with mods.
- [ ] I recall current UI being sometimes heavier to render on some configurations than actual game. See https://github.com/colobot/colobot/issues/895. This question is potentially outdated with new renderer.

The reason for moving UI layout to external files is to make it possible for engine to handle other games as well (Blupimania, BuzzingCars) without baking into code all UI layouts and choose between them at compile time or by `switch()case{}` construction inside. Each game would just have its own layout file, processed at runtime.
New UI system should also - when possible - solve some of existing UI-related issues https://github.com/colobot/colobot/labels/ui.

Originally we've planned to use [CEGUI](https://github.com/cegui/cegui) library. It allows you to define UI with XML files and delivers CEED - graphical editor to create UI layout. 

Another popular library is [Dear ImGui](https://github.com/ocornut/imgui), however it seems more suitable for quick debug menus and small applications than complex game UI. Also I couldn't find anything about defining layout in external files.

**Comment by hexagonrecursion** [comment]
> * Definable by external files, instead of being hardcoded in engine. Possible to enhance/replace with mods.

Beware: this will run into similar issues as [GTK style sheet theming](https://blogs.gnome.org/tbernard/2018/10/15/restyling-apps-at-scale/):

1. Adding UI breaks mods
   1. A mod replaces an xml that defines e.g the pause menu
   2. A new version of colobot adds a new button to the pause menu
   3. Now the mod is broken because the button disappears if you instal the mod
1. Reorganizing UI breaks mods
   1. A mod replaces an xml that defines e.g the settings screen
   2. A new version of colobot [replaces a slider with an EditValue](https://github.com/colobot/colobot/issues/723)
   3. Now the mod has a slider that does nothing and the EditValue is missing

Using xml files to define UI and allowing mods to replace them creates a huge API surface that we have to support. Adding any new UI will break the API. We could try to avoid breaking the API when reorganizing UI, but (1) writing code on the C++ side to support both the new xml and the old xml would substantially increase complexity (2) we will accidentally break the mods anyway - we do not have the manpower to check every UI element across dozens of mods in case one is broken.

**Reply by tomangelo2** [reply]
Yeah, that might be a problem. 
We could make it tied to specific engine release, similar to how other games works (game gets an update - mod creators have to update it to make it compatible with core objects and API changes). Running UI mods created for different release would be unsupported, and that's something mod creators would have to acknowledge when making such mod. The game would display a warning when loading UI mod made for different release.

Now for theming. Let's say there is mod A that is a general theme overhaul and mod B, that adds new robot with it's own UI. Are there any tools that would prevent UI from B to work fine with default theme, but glitch here and there with A theme? Either we have to limit what you can do (reducing possible combinations that we'd have to support), or we'd eventually run into GTK theming issues.

Maybe instead of defining whole UI with XML/JSON we'd just implement some "core" UI with "fixed" look (customizable through changing textures, or simple controls), with game-specific windows/areas as widgets defined in external files, with limited API?

**Comment by hexagonrecursion** [comment]
> mod B, that adds new robot with it's own UI

How? Mods can't add custom C++ code and we probably don't want to change that. Every button requires an event handler to be written for it. Most buttons also require code to show/hide or enable/disable. Every checkbox, every dropdown, every text field requires some code to initially set it's value and more code to read it later and do something. Giving mods an ability to add a custom xml UI is useless if none of the UI elements actually do anything.

**Reply by tomangelo2** [reply]
I was thinking about how BuilderBots have their own UI, that reuses part of human UI. There are some properties of object that would in some cases be useful to display.
First idea would be BuzzingCars cars with speedometer, or display it's health points as car gauge instead of vertical bar. Or some flying units that would have gauges for yaw/roll/pitch, or maybe a robot/building would use 2 or more resources from their slots and we'd like to display second resource as well.

**Reply by hexagonrecursion** [reply]
I still do not understand.
1. BuilderBot UI adds an entirely new button - the button to show/hide the list of buildings. This button needs an event handler. How can a mod add an event handler if it can't add C++ code?
2. "BuzzingCars cars with speedometer". To add that the mod would have to add code to compute the speed and update the value that the speedometer displays. How?
3. same for car gauge, yaw/roll/pitch and multiple resources

I do agree that it would be neat if mods could do that. My point is that an xml is not sufficient - it can define how things look and where they are positioned, but it can't define how they function.

**Comment by hexagonrecursion** [comment]
> Another popular library is [Dear ImGui](https://github.com/ocornut/imgui), however it seems more suitable for quick debug menus and small applications than complex game UI

Could you please elaborate? I am aware of at least two games ([Cataclysm-DDA](https://github.com/CleverRaven/Cataclysm-DDA) and [Pioneer](https://github.com/pioneerspacesim/pioneer)) that successfully use ImGui

> The game would display a warning when loading UI mod made for different release.

I doubt that this would be effective.
1. Alice writes a UI mod
2. Every new version of the game displays a warning when loading Alice's mod just in case it is incompatible, *but* in practice everything works
3. Alice learns that to make her mod compatible with the new version of the game she just has to increment the version number
4. When the mod does break something, the warning will be missed

**Reply by tomangelo2** [reply]
I've looked on ImGui readme, which says:
```
Dear ImGui is designed to enable fast iterations and to empower programmers to create content creation tools and visualization / debug tools (as opposed to UI for the average end-user). It favors simplicity and productivity toward this goal and lacks certain features commonly found in more high-level libraries. Among other things, full internationalization (right-to-left text, bidirectional text, text shaping etc.) and accessibility features are not supported.
```
Maybe I've misunderstood this paragraph, I thought that since it's designed for quick creation of simple windows it might not be suitable for games with more complex controls.

As for incompatibility warning: this might happen on any mod. Let's say we're changing how UV mapping of terrain works, or change colors scale from 3 0-1 floats to one RGB hex code. If mod/level creator won't check out and just blindly marks it as compatible, it will also break.
Of course, making such compatibility breaking changes shouldn't happen very often (at least since 1.0.0, since API-breaking changes should increase major number), but UI will in fact be more vulnerable to such changes than level interpreter or 3D model loader.

The main goal for external UI definition is to tie it with game, not engine. Modding would be a nice feature, but if it's going to create more issues than profits, then I'm fine with dropping that part.

**Reply by hexagonrecursion** [reply]
* full internationalization (right-to-left text, bidirectional text, text shaping etc.) not supported
* accessibility features not supported

That sounds like a dealbreaker. Does CEGUI support those?

**Reply by tomangelo2** [reply]
http://cegui.org/forum/viewtopic.php?t=3929&p=18193 <- right-to-left and bidirectional text
http://cegui.org.uk/wiki/Formatting_Tags_in_CEGUI <- text formatting
However I can't find much about translations support or accessibility features. Also their forum and wiki seems to not be updated recently (they moved to Discord for discussions). Either I'm not searching efficiently enough, or these might be a deal-breaker for CEGUI as well.

**Comment by hexagonrecursion** [comment]
Let's first write down our requirements for a GUI framework. Requirements established so far:
1. The framework supports at least Linux(required: x86_64; nice to have: ARM64), macOS(ARM64), Windows(required: x86_64; nice to have: ARM64). ~Do we want 32 bit too? What about Linux on arm?~ - answered
2. The framework is actively maintained
3. The framework has comprehensive up-to-date documentation
15. The framework has a test suite with a decent code coverage percentage
4. The framework supports screen readers on all the above platforms. I am not sure if we have the resources to make the game accessible to blind people, but we could at least improve accissibility for people with poor eye sight
5. Color blind support: when comparing different frameworks we should take into consideration how they enable us to make the game accessible to color blind people
6. The framework supports bidirectional text (e.g Hebrew comments in a CBot program)
7. The framework supports Unicode
8. The framework supports Wayland on Linux (unless it takes input from the user exclusively through SDL and uses SDL or OpenGL to draw the UI). Note: I did not mention X11 because I assume that anything claiming to support Linux supports it
9. The framework supports SDL2 and SDL3 support (if relevant). SDL2 because that is what we currently use. SDL3 in case we want to upgrade in the future
11. Optional: it would be nice if it had some kind of API reflection or a machine readable description of it's API that would enable us to automatically generate CBot bindings for it to enable UI mods
12. Need to be able to implement a CBot editor with:
    1. Syntax highlighting
    2. Search https://github.com/colobot/colobot/issues/276
    3. Undo/redo
13. Need to be able to implement a SatCom viewer with: pictures, fonts, colours, clickable text that navigates to other SatCom pages
18. Nice to have: touchscreen support

Questions that are still unanswered:

1. Do we need Vulcan support? Do we need DirectX support? Or is OpenGL enough?
2. ~Do we want to be able to link it statically? Is a framework that can only be dynamically linked ok?~ answered
3. ~Do we want to be able to link it dynamically? Is a framework that can only be statically linked ok?~ answered
6. Do we need controller support?
7. Do we need joystick support?
8. Does the GUI framework need to be packaged in repositories? Which repositories?
9. Do we want to be able to add the GUI framework as a submodule? If so, does it need to have a CMakeLists.txt? It appears that the approach we currently use only works if the submodule has CMakeLists.txt.
10. Did I forget to add anything?

**Reply by tomangelo2** [reply]
Linux (and possible Windows as well) on ARM - I think it would be nice to have, so if ARM platform would become more popular we wouldn't be left behind.
32-bit support - I think we don't need it.  By that point Pentium IV era computers might have issues to run Colobot due to OpenGL 3.3 requirement. And there are barely anyone running 32-bit OS on x86_64 CPUs. 

For static/dynamic linking - I don't have any strong opinion on any of these options.

If possible - touchscreens support could be useful if we'd port the game to mobile platforms.

And since accessibility is also mentioned, I'll link existing issue about it https://github.com/colobot/colobot/issues/1451. UI supporting colorblind mode or captions for sound effects for hearing impaired would also be a nice to have.

**Reply by hexagonrecursion** [reply]
> UI supporting colorblind mode or captions for sound effects for hearing impaired would also be a nice to have.

Correct me if I'm wrong, but this sounds like something we would have to implement on top of the UI framework rather than a feature of the framework itself. For captions we just need a label. I assume every UI framework provides that. For colorblind support we need to be able to change colors of UI elements - I assume evey UI framework can do that.

> Linux (and possible Windows as well) on ARM - I think it would be nice to have, so if ARM platform would become more popular we wouldn't be left behind.
> 32-bit support - I think we don't need it. By that point Pentium IV era computers might have issues to run Colobot due to OpenGL 3.3 requirement. And there are barely anyone running 32-bit OS on x86_64 CPUs.
> If possible - touchscreens support could be useful if we'd port the game to mobile platforms.

Thanks. I updated the requirement list

**Reply by tomangelo2** [reply]
I'm not sure how different UIs are handling colorblind modes. If they all rely on same technique, like recoloring textures by shader like it currently works for textures, then it's only a matter if some handle it better than others. If there is more than one way to do it, then it's another surface we can compare them.

**Reply by hexagonrecursion** [reply]
I have more questions about the requirements. I have edited the questions into my comment above.
